src/express/commands/milestone.js

```diff
@@ -0,0 +1,400 @@
+  let tasks = []
+  for (let i = 0; i < ips.length; i++) {
+    if (i == 0) {
+      // remove all other peers from 1st node
+      tasks.push(removePeers(ips[i], enodes.slice(1)))
+    } else {
+      // remove 1st node from all peers
+      tasks.push(removePeers(ips[i], [enodes[0]]))
+    }
+  }
+
+  await Promise.all(tasks).then((values) => {
+    if (values.includes(false)) {
+      console.log('📍Unable to remove peers for creating cluster, exiting')
+      return false
+    }
+  }).catch(error => {
+    return false
+  })
+
+  return true
+}
+
+async function rejoinClusters(ips, enodes) {
+  let tasks = []
+  for (let i = 0; i < ips.length; i++) {
+    if (i == 0) {
+      // add all other peers in 1st node
+      tasks.push(addPeers(ips[i], enodes.slice(1)))
+    } else {
+      // add 1st node in all peers
+      tasks.push(addPeers(ips[i], [enodes[0]]))
+    }
+  }
+
+  await Promise.all(tasks).then((values) => {
+    if (values.includes(false)) {
+      console.log('📍Unable to add peers, exiting')
+      return false
+    }
+  }).catch(error => {
+    console.log(`📍Unable to add peers, error: ${error}`)
+    return false
+  })
+}
+
+async function getEnode(user, host) {
+  const ip = `${user}@${host}`
+  let command = '~/go/bin/bor attach ~/.bor/data/bor.ipc --exec admin.nodeInfo.enode'
+  try {
+    const fullEnode = await runSshCommandWithReturn(ip, command, maxRetries)
+    let enode = String(fullEnode).split('@')[0].slice(1); // remove the local ip from the enode
+    if (enode.length != 136) { // prefix "enode://" + 128 hex values for enode itself
+      return ''
+    }
+    enode += "@" + host + ":30303" // assuming that p2p port is opened on 30303 
+    return enode
+  } catch (error) {
+    console.log(`📍Unable to query enode. Error: ${error}`)
+  }
+
+  return ''
+}
+
+function validateNumberOfPeers(peers) {
+  let recreate = false
+
+  // 1st node should have 0 peers
+  if (peers[0] > 0) {
+    console.log('📍Unexpected peer length received for 1st cluster, retrying. expected: 0, received:', peers[0])
+    recreate = true
+  }
+
+  // Remaining nodes should have total-1 peers
+  for (let i = 1; i < peers.length; i++) {
+    if (peers[i] != (peers.length - 2)) {
+      console.log('📍Unexpected peer length received for 2nd cluster, retrying')
+      recreate = true
+      break
+    }
+  }
+
+  return recreate
+}
+
+export async function milestone() {
+  // NOTE: Make sure bor branch has logic for hardcoded primary validator
+  
+  require('dotenv').config({ path: `${process.cwd()}/.env` })
+  const devnetType =
+    process.env.TF_VAR_DOCKERIZED === 'yes' ? 'docker' : 'remote'
+
+  const doc = await loadDevnetConfig(devnetType)
+
+  const borUsers = splitToArray(doc['devnetBorUsers'].toString())
+  const borHosts = splitToArray(doc['devnetBorHosts'].toString())
+
+  // Check for number of validators
+  if (doc['numOfValidators'].length < 4) {
+    console.log('📍Cannot run milestone tests on less than 4 validator nodes')
```

Comment:
can you explain why we need >=4 vals?

Reply:
Just for simplicity of the tests. If the number of validators are 3 (or less), we won't get group with majority votes while creating clusters. As it's going to be used only internally, I added this check.